ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java

```diff
@@ -96,74 +101,237 @@ public void close() {
-                } catch (Exception e) {
-                    log.error("Error in refreshIngestionAuthToken. " + e.getMessage(), e);
-                    if (timer != null) {
-                        timer.schedule(new RefreshIngestionAuthTokenTask(), refreshTimeOnFailure);
+                    return null;
+                }, "ResourceManager.refreshIngestionAuthToken");
+            } catch (Exception e) {
+                log.error("Error in refreshIngestionAuthToken. " + e.getMessage(), e);
+                scheduleRefreshIngestionAuthTokenTask(refreshTimeOnFailure);
+            }
+        }
+
+        private void refreshIngestionAuthToken() throws IngestionClientException, IngestionServiceException {
+            if (ingestionAuthTokenLock.writeLock().tryLock()) {
+                try {
+                    log.info("Refreshing Ingestion Auth Token");
+                    Retry retry = Retry.of("get Ingestion Auth Token resources", retryConfig);
+                    CheckedFunction0<KustoOperationResult> retryExecute = Retry.decorateCheckedSupplier(retry,
+                            () -> client.execute(Commands.IDENTITY_GET_COMMAND));
+                    KustoOperationResult identityTokenResult = retryExecute.apply();
+                    if (identityTokenResult != null
+                            && identityTokenResult.hasNext()
+                            && !identityTokenResult.getResultTables().isEmpty()) {
+                        KustoResultSetTable resultTable = identityTokenResult.next();
+                        resultTable.next();
+                        identityToken = resultTable.getString(0);
                     }
+                    refreshedAtLeastOnce = true;
+                    ingestionAuthTokenRefreshedOnce.signalAll();
+                } catch (DataServiceException e) {
+                    throw new IngestionServiceException(e.getIngestionSource(), "Error refreshing IngestionAuthToken. " + e.getMessage(), e);
+                } catch (DataClientException e) {
+                    throw new IngestionClientException(e.getIngestionSource(), "Error refreshing IngestionAuthToken. " + e.getMessage(), e);
+                } catch (Throwable e) {
+                    throw new IngestionClientException(e.getMessage(), e);
+                } finally {
+                    ingestionAuthTokenLock.writeLock().unlock();
                 }
             }
         }
+    }
 
-        timer.schedule(new RefreshIngestionAuthTokenTask(), 0);
-        timer.schedule(new RefreshIngestionResourcesTask(), 0);
+    private void init() {
+        scheduleRefreshIngestionResourcesTask(0L);
+        scheduleRefreshIngestionAuthTokenTask(0L);
+    }
+
+    // If we combined these 2 methods, ensuring we're using distinct synchronized locks for both would be inelegant
+    private synchronized void scheduleRefreshIngestionResourcesTask(Long delay) {
```

Comment:
Making the method `synchronized` ensures the task is never `null` and is always scheduled.